File: backend/src/topic_scoring.py
DIFFICULTY_WEIGHTS = {"Easy": 1, "Medium": 2, "Hard": 3}

TOPIC_STATUS_THRESHOLDS = {
    "weak_max": 59,
    "medium_max": 79,
}

_BREAKDOWN_KEYS = {"Easy": "easy", "Medium": "medium", "Hard": "hard"}
_LEGACY_TOPIC_ALIASES = {"General": "Uncategorized"}


def _safe_int(value, default=0):
    try:
        if value is None:
            return default
        return int(value)
    except (TypeError, ValueError):
        return default


def _normalize_topic_key(topic):
    return _LEGACY_TOPIC_ALIASES.get(topic, topic)


def _difficulty_weight(difficulty):
    return DIFFICULTY_WEIGHTS.get(difficulty, 2)


def _topic_status(score):
    if score <= TOPIC_STATUS_THRESHOLDS["weak_max"]:
        return "weak"
    if score <= TOPIC_STATUS_THRESHOLDS["medium_max"]:
        return "medium"
    return "strong"


def _merge_matrix(matrix):
    merged = {}
    if not isinstance(matrix, dict):
        return merged

    for topic_key, difficulties in matrix.items():
        if not isinstance(difficulties, dict):
            continue
        canonical = _normalize_topic_key(topic_key)
        topic_cells = merged.setdefault(canonical, {})
        for difficulty, cell in difficulties.items():
            if not isinstance(cell, dict):
                continue
            bucket = topic_cells.setdefault(difficulty, {"correct": 0, "total": 0})
            bucket["correct"] += _safe_int(cell.get("correct"))
            bucket["total"] += _safe_int(cell.get("total"))
    return merged


def compute_topic_scores(matrix):
    merged = _merge_matrix(matrix)
    topics = []

    for topic, difficulties in merged.items():
        weighted_correct = 0
        weighted_total = 0
        correct_count = 0
        total_answered = 0
        difficulty_breakdown = {
            key: {"correct": 0, "total": 0, "score": None}
            for key in ("easy", "medium", "hard")
        }

        for difficulty, cell in difficulties.items():
            if not isinstance(cell, dict):
                continue
            correct_d = _safe_int(cell.get("correct"))
            total_d = _safe_int(cell.get("total"))
            weight = _difficulty_weight(difficulty)
            weighted_correct += weight * correct_d
            weighted_total += weight * total_d
            correct_count += correct_d
            total_answered += total_d

            breakdown_key = _BREAKDOWN_KEYS.get(difficulty)
            if breakdown_key is not None:
                difficulty_breakdown[breakdown_key]["correct"] += correct_d
                difficulty_breakdown[breakdown_key]["total"] += total_d

        for breakdown in difficulty_breakdown.values():
            total_d = breakdown["total"]
            if total_d > 0:
                breakdown["score"] = round((breakdown["correct"] / total_d) * 100)

        if weighted_total == 0:
            continue

        score = round((weighted_correct / weighted_total) * 100)
        topics.append(
            {
                "topic": topic,
                "score": score,
                "status": _topic_status(score),
                "total_answered": total_answered,
                "correct_count": correct_count,
                "wrong_count": total_answered - correct_count,
                "weighted_correct": weighted_correct,
                "weighted_total": weighted_total,
                "difficulty_breakdown": difficulty_breakdown,
            }
        )

    topics.sort(key=lambda item: (item["score"], item["topic"]))
    return topics

File: frontend/src/utils/topicScoring.js
const DIFFICULTY_WEIGHTS = { Easy: 1, Medium: 2, Hard: 3 }

const TOPIC_STATUS_THRESHOLDS = {
  weak_max: 59,
  medium_max: 79,
}

const BREAKDOWN_KEYS = { Easy: 'easy', Medium: 'medium', Hard: 'hard' }
const LEGACY_TOPIC_ALIASES = { General: 'Uncategorized' }

function safeInt(value, defaultValue = 0) {
  const parsed = Number(value)
  if (!Number.isFinite(parsed)) return defaultValue
  return Math.trunc(parsed)
}

function normalizeTopicKey(topic) {
  return LEGACY_TOPIC_ALIASES[topic] ?? topic
}

function difficultyWeight(difficulty) {
  return DIFFICULTY_WEIGHTS[difficulty] ?? 2
}

function topicStatus(score) {
  if (score <= TOPIC_STATUS_THRESHOLDS.weak_max) return 'weak'
  if (score <= TOPIC_STATUS_THRESHOLDS.medium_max) return 'medium'
  return 'strong'
}

function mergeMatrix(matrix) {
  const merged = {}
  if (!matrix || typeof matrix !== 'object') return merged

  for (const [topicKey, difficulties] of Object.entries(matrix)) {
    if (!difficulties || typeof difficulties !== 'object') continue
    const canonical = normalizeTopicKey(topicKey)
    if (!merged[canonical]) merged[canonical] = {}

    for (const [difficulty, cell] of Object.entries(difficulties)) {
      if (!cell || typeof cell !== 'object') continue
      if (!merged[canonical][difficulty]) {
        merged[canonical][difficulty] = { correct: 0, total: 0 }
      }
      merged[canonical][difficulty].correct += safeInt(cell.correct)
      merged[canonical][difficulty].total += safeInt(cell.total)
    }
  }
  return merged
}

export function computeTopicScores(matrix) {
  const merged = mergeMatrix(matrix)
  const topics = []

  for (const [topic, difficulties] of Object.entries(merged)) {
    let weightedCorrect = 0
    let weightedTotal = 0
    let correctCount = 0
    let totalAnswered = 0
    const difficultyBreakdown = {
      easy: { correct: 0, total: 0, score: null },
      medium: { correct: 0, total: 0, score: null },
      hard: { correct: 0, total: 0, score: null },
    }

    for (const [difficulty, cell] of Object.entries(difficulties)) {
      if (!cell || typeof cell !== 'object') continue
      const correctD = safeInt(cell.correct)
      const totalD = safeInt(cell.total)
      const weight = difficultyWeight(difficulty)
      weightedCorrect += weight * correctD
      weightedTotal += weight * totalD
      correctCount += correctD
      totalAnswered += totalD

      const breakdownKey = BREAKDOWN_KEYS[difficulty]
      if (breakdownKey) {
        difficultyBreakdown[breakdownKey].correct += correctD
        difficultyBreakdown[breakdownKey].total += totalD
      }
    }

    for (const breakdown of Object.values(difficultyBreakdown)) {
      if (breakdown.total > 0) {
        breakdown.score = Math.round((breakdown.correct / breakdown.total) * 100)
      }
    }

    if (weightedTotal === 0) continue

    const score = Math.round((weightedCorrect / weightedTotal) * 100)
    topics.push({
      topic,
      score,
      status: topicStatus(score),
      total_answered: totalAnswered,
      correct_count: correctCount,
      wrong_count: totalAnswered - correctCount,
      weighted_correct: weightedCorrect,
      weighted_total: weightedTotal,
      difficulty_breakdown: difficultyBreakdown,
    })
  }

  topics.sort((a, b) => a.score - b.score || a.topic.localeCompare(b.topic))
  return topics
}

export function resolveProgressTopics(payload) {
  if (Array.isArray(payload?.topics)) {
    return payload.topics
  }
  const matrix = payload?.matrix
  if (matrix && typeof matrix === 'object') {
    return computeTopicScores(matrix)
  }
  return []
}
